### ISSUE TITLE
[R][Documentation] Document add_filename on open_dataset help page

### ISSUE THREAD

**Issue by amoeba** [apache/arrow#38771]
### Describe the enhancement requested

In https://github.com/apache/arrow/issues/38724, it was pointed out that `add_filename` could be easier to find. Documenting it on the open_dataset help page would be a good start and maybe whoever does that might find other places it could be documented better too.

### Component(s)

Documentation, R

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.

**Comment by PietrH** [comment]
This function is currently internal, is it supposed to be exported? Same for `cast()`

**Comment by thisisnic** [comment]
Nah, they're a bit unusual in that they work within arrow queries rather than as standalone functions which means that we don't want to export them

**Comment by PietrH** [comment]
Alright, what would be the best way to use them within a package ?

**Comment by thisisnic** [comment]
Easiest way to explain was to make this PR - see the write-up in #51392 - thanks for the nudge!
